pkg/engine/engine.go

```diff
@@ -255,19 +291,55 @@ func (e Engine) renderWithReferences(tpls, referenceTpls map[string]renderable)
 		}
 		// At render time, add information about the template that is being rendered.
 		vals := tpls[filename].vals
-		vals["Template"] = chartutil.Values{"Name": filename, "BasePath": tpls[filename].basePath}
-		var buf strings.Builder
-		if err := t.ExecuteTemplate(&buf, filename, vals); err != nil {
-			return map[string]string{}, cleanupExecError(filename, err)
+		if vals["Template"] == nil {
+			vals["Template"] = chartutil.Values{"Name": filename, "BasePath": tpls[filename].basePath}
 		}
+		if toplevel {
+			go doRenderAsync(renderResults, filename, t, vals)
+		} else {
+			r, renderError := doRender(filename, t, vals)
+			if renderError != nil {
+				return map[string]string{}, cleanupExecError(filename, renderError)
+			}
+			rendered[filename] = r
+		}
+	}
+	if renderResults != nil {
+		for _, filename := range keys {
+			if strings.HasPrefix(path.Base(filename), "_") {
+				continue
+			}
+			renderResult := <-renderResults
+			if renderResult.error != nil {
+				return map[string]string{}, cleanupExecError(renderResult.filename, renderResult.error)
+			}
+			rendered[renderResult.filename] = renderResult.rendered
+		}
+	}
+	return rendered, nil
+}
+
+type RenderResults struct {
```

Comment:
Unless this is something that is returned in a function signature, this should be marked private to the package using `renderResults` instead of `RenderResults` as the name.